addons/binding/org.openhab.binding.regoheatpump/src/main/java/org/openhab/binding/regoheatpump/handler/Rego6xxHeatPumpHandler.java

```diff
@@ -0,0 +1,405 @@
+                readFromFrontPanel(channelIID, (short) 0x0016);
+                break;
+
+            default:
+                readFromSystemRegister(channelIID);
+                break;
+        }
+    }
+
+    private Collection<String> linkedChannels() {
+        return thing.getChannels().stream().map(Channel::getUID).map(ChannelUID::getId).filter(this::isLinked)
+                .collect(Collectors.toList());
+    }
+
+    private void refresh() {
+        for (final String channelIID : linkedChannels()) {
+            if (Thread.interrupted()) {
+                break;
+            }
+
+            processChannelRequest(channelIID);
+
+            if (thing.getStatus() != ThingStatus.ONLINE) {
+                break;
+            }
+        }
+    }
+
+    private void readLastErrorType() {
+        readLastError(CHANNEL_LAST_ERROR_TYPE, e -> new StringType(Byte.toString(e.error())));
+    }
+
+    private void readLastErrorTimestamp() {
+        readLastError(CHANNEL_LAST_ERROR_TIMESTAMP, e -> new DateTimeType(e.timestamp()));
+    }
+
+    private void readLastError(final String channelIID, final Function<ErrorLine, State> converter) {
+        executeCommandAndUpdateState(channelIID, CommandFactory.createReadLastErrorCommand(),
+                ResponseParserFactory.ErrorLine, e -> {
+                    return e == null ? UnDefType.NULL : converter.apply(e);
+                });
+    }
+
+    private void readFromFrontPanel(final String channelIID, short address) {
+        final byte[] command = CommandFactory.createReadFromFrontPanelCommand(address);
+        executeCommandAndUpdateState(channelIID, command, ResponseParserFactory.Short, DecimalType::new);
+    }
+
+    private void readFromSystemRegister(final String channelIID) {
+        final RegoRegisterMapper.Channel channel = mapper.map(channelIID);
+        if (channel != null) {
+            final byte[] command = CommandFactory.createReadFromSystemRegisterCommand(channel.address());
+            executeCommandAndUpdateState(channelIID, command, ResponseParserFactory.Short, channel::convert);
+        } else {
+            logger.debug("Unsupported channel {}", channelIID);
+        }
+    }
+
+    private synchronized <T> void executeCommandAndUpdateState(final String channelIID, final byte[] command,
+            final ResponseParser<T> parser, Function<T, State> converter) {
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("Reading value for channel '{}' ...", channelIID);
+        }
+
+        try {
+
+            final T result = executeCommandWithRetry(channelIID, command, parser, 5);
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("Got value for '{}' = {}", channelIID, result);
+            }
+
+            updateState(channelIID, converter.apply(result));
+
+        } catch (IOException e) {
+
+            logger.warn("Accessing value for channel '" + channelIID + "' failed.", e);
```

Comment:
I used this only to get full exception back stack - a quick google and found this - https://www.slf4j.org/faq.html#paramException - updated :)